pkg/vsphere/extraconfig/decode.go

```diff
@@ -0,0 +1,278 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extraconfig
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// fromString converts string representation of a basic type to basic type
+func fromString(field reflect.Value, value string) reflect.Value {
+	switch field.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		s, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			log.Errorf("Failed to convert: %s", err.Error())
+			return field
+		}
+		return reflect.ValueOf(int(s))
+
+	case reflect.Bool:
+		s, err := strconv.ParseBool(value)
+		if err != nil {
+			log.Errorf("Failed to convert: %s", err.Error())
+			return field
+		}
+		return reflect.ValueOf(s)
+
+	case reflect.String:
+		return reflect.ValueOf(value)
+
+	case reflect.Float32, reflect.Float64:
+		s, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			log.Errorf("Failed to convert: %s", err.Error())
+			return field
+		}
+		return reflect.ValueOf(s)
+
+	}
+	log.Debugf("Invalid Kind: %s", field.Kind())
+
+	return field
+}
+
+// decodeWithPrefix convert []types.BaseOptionValue to type dest
+func decodeWithPrefix(kv map[string]string, dest interface{}, prefix string) interface{} {
+	// value representing the run-time data
+	value := reflect.ValueOf(dest).Elem()
+	log.Debugf("Value: %#v", value)
+
+	// determine the kind as it changes how we get underlying data
+	switch value.Kind() {
+	case reflect.Invalid:
+		log.Errorf("Invalid Kind: %#v", value)
+		return nil
+
+	case reflect.Struct:
+		// iterate through every struct type field
+		for i := 0; i < value.NumField(); i++ {
+			// get the underlying struct field
+			structField := value.Type().Field(i)
+			//skip unexported fields
+			if structField.PkgPath != "" {
+				log.Debugf("Skipping %s (not exported)", structField.Name)
+				continue
+			}
+
+			// get the annotations
+			tags := structField.Tag
+			log.Debugf("Tags: %#v", tags)
+
+			// do we have DefaultTagName?
+			if tags.Get(DefaultTagName) == "" {
+				log.Debugf("Skipping %s (not vic)", structField.Name)
+				continue
+			}
+
+			// get the scopes
+			scopes := strings.Split(tags.Get("scope"), ",")
+			log.Debugf("Scopes: %#v", scopes)
+
+			// get the keys and split properties from it
+			keys := strings.Split(tags.Get("key"), ",")
+			key, properties := keys[0], keys[1:]
+			log.Debugf("Keys: %#v Properties: %#v", key, properties)
+
+			// returns the i'th field of the struct i
+			field := value.Field(i)
+
+			// process properties
+			if len(properties) > 0 {
+				// do not recurse into nested type
+				if properties[0] == "omitnested" {
+					log.Debugf("Skipping %s (omitnested)", key)
```

Comment:
can we tell if there are values that would be unpacked under this type if not omitted? Can we issue a warning if that's the case?